### PR TITLE
🐛 Performs KCP version check on control plane update

### LINKED ISSUE #2578: KCP: validate version changes are no further than one minor version
**User Story**

As an operator, I would like to get immediate feedback when I try to ask the KCP to perform a multi-version upgrade.

**Detailed Description**

Pulling this out of #2489: if a KCP is pointing to version 1.16.x, it should reject updates that ask to upgrade the control plane directly to 1.18.x.

**Anything else you would like to add:**

I'm not sure what the semantics should be around downgrades – do we deny those outright?

/kind feature

### PR DESCRIPTION
**Pull request by srm09**
**What this PR does / why we need it**:
This PR is needed to disallow users from downgrading the control plane versions as well as skipping minor versions during upgrades.

**Which issue(s) this PR fixes**:
Fixes #2578

### CODE REVIEW

**Comment by srm09** [comment]
@vincepri Does it make sense to add logic for major versions?

**Comment by neolit123** [comment]
hi, would there be a way to override the behavior of this webhook and not error out on downgrade, minor skip?

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"strings"
 
 	"github.com/coredns/corefile-migration/migration"
+	"github.com/pkg/errors"
 	kubeadmv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
```
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:28]: Remove this line?

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 2]
```diff
@@ -384,6 +386,52 @@ func (in *KubeadmControlPlane) validateEtcd(prev *KubeadmControlPlane) (allErrs
...
+		return allErrs
+	}
+
+	if toVersion.LT(fromVersion) {
+		allErrs = append(allErrs,
+			field.Forbidden(
+				field.NewPath("spec", "version"),
+				fmt.Sprintf("cannot downgrade control plane from %s to %s", previousVersion, in.Spec.Version),
+			),
+		)
```
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:418]: Should we allow rollback though? This seems it might impact the ability to stop and rollback an upgrade.

I'd leave this out for now and only validate the upgrade path, where we need to check that you can only upgrade one minor version

cc @detiber

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 3]
```diff
@@ -384,6 +386,52 @@ func (in *KubeadmControlPlane) validateEtcd(prev *KubeadmControlPlane) (allErrs
...
+	} else {
+		if toVersion.Major - fromVersion.Major > 1 ||
+			(toVersion.Major - fromVersion.Major == 1 && toVersion.Minor != 0) ||
+			(toVersion.Major == fromVersion.Major && toVersion.Minor - fromVersion.Minor > 1) {
```
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:422]: Can we add comment around this? It's a little hard to follow what we're doing here.

In alternative to comparing one by one, we could create a "ceiling version" from the information we have in the `fromVersion`, and then make sure the `toVersion` is less or equal to the `ceilingVersion`, wdyt?

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go (hunk 2 above)
Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:418]: Alright, I will remove the downgrade part for now.

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go (hunk 3 above)
Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:422]: Defining a ceiling version might be harder than expected. Mainly because, say for instance, the current KCP version is `1.17.4`. Then, we could decide that the ceiling version is `1.18.x`.
But it could also be the case, that the next version actually released was a major version bump, say `2.0.0`.  So, (to tackle this scenario), if we keep the ceiling as `2.0.0`, then ideally, an update could skip over multiple minor versions which is what we wanna avoid.
Does it make sense?

Comment (vincepri) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:422]: Yeah, although the plan for now is to stay on Kubernetes 1.x, if we ever go down the route of a major release, there is lots of changes that need to happen

Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:422]: Honestly I could remove it and keep the logic simple enough.

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go (hunk 1 above)
Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:28]: This import is required for the `errors.Wrapf()` call. Are you suggesting I get rid off that, and just return a plain error with a useful message.

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 4]
```diff
@@ -384,6 +386,47 @@ func (in *KubeadmControlPlane) validateEtcd(prev *KubeadmControlPlane) (allErrs
...
+	ceilVersion := semver.Version{
+		Major: fromVersion.Major,
+		Minor: fromVersion.Minor + 2,
+		Patch: 0,
+	}
+	if toVersion.GTE(ceilVersion) {
+		allErrs = append(allErrs,
+			field.Forbidden(
+				field.NewPath("spec", "version"),
+				fmt.Sprintf("cannot skip minor versions of kubeadmcontrolplane from %s to %s", previousVersion, in.Spec.Version),
```
Comment (vincepri) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:422]:
```suggestion
				fmt.Sprintf("cannot update Kubernetes version from %s to %s", previousVersion, in.Spec.Version),
```

We should probably add some more context here, do we have a link handy to Kubernetes version upgrade policy? 

cc @fabriziopandini

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 5]
```diff
@@ -384,6 +386,47 @@ func (in *KubeadmControlPlane) validateEtcd(prev *KubeadmControlPlane) (allErrs
...
+		)
+		return allErrs
+	}
+
+	toVersion, err := util.ParseMajorMinorPatch(in.Spec.Version)
+	if err != nil {
+		allErrs = append(allErrs,
+			field.InternalError(
+				field.NewPath("spec", "version"),
+				errors.Wrapf(err, "failed to parse current kubeadmcontrolplane version: %s", previousVersion),
```
[the remark this answers is not in the document]
Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:406]: We have added the faulty version in the error string, that should give enough information as to what version string is incorrect.

[the remark this answers is not in the document]
Comment (ncdc) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:406]: True, but in the previous check (line 395), you're looking at what's in etcd (which should never fail), and it makes sense to me to call that "current". The value you're examining here is the proposed updated version. If we can be explicit for our users, that's usually best, so they don't have to do any guessing. WDYT?

[the remark this answers is not in the document]
Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:406]: Sold! I will make the change :)

**Review by cpanato — APPROVED** [review]
/lgtm

/hold to wait for @ncdc since he made a previous comment to change

**Comment by ncdc** [comment]
LGTM, thanks!

/hold cancel

File: controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go [hunk 6]
```diff
@@ -384,6 +386,47 @@ func (in *KubeadmControlPlane) validateEtcd(prev *KubeadmControlPlane) (allErrs
...
+			field.InternalError(
+				field.NewPath("spec", "version"),
+				errors.Wrapf(err, "failed to parse updated kubeadmcontrolplane version: %s", in.Spec.Version),
+			),
+		)
+		return allErrs
+	}
+
+	// since upgrades to the next minor version are allowed, irrespective of the patch version
+	ceilVersion := semver.Version{
```
Comment (benmoss) [inline controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:413]: This is not the ceiling but actually the first disallowed version

Comment (srm09) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:413]: Should I make a tiny PR to address this @benmoss ?

Comment (benmoss) [reply controlplane/kubeadm/api/v1alpha3/kubeadm_control_plane_webhook.go:413]: Meh, I don't think you need to bother
